python/lsst/resources/s3utils.py

```diff
@@ -117,25 +114,29 @@ class _TooManyRequestsError(Exception):
 max_retry_time = 60
 
 
-def clean_test_environment(testcase: TestCase) -> None:
-    """Clear S3_ENDPOINT_URL then restore it at the end of a test.
-
-    Parameters
-    ----------
-    testcase: `unittest.TestCase`
-        Reference to the test being run; used to add a cleanup function.
+@contextmanager
+def clean_test_environment_for_s3() -> Iterator[None]:
+    """Reset S3 environment to ensure that unit tests with a mock S3 can't
+    accidentally reference real infrastructure
     """
-    endpoint = os.environ.get("S3_ENDPOINT_URL")
-
-    if not endpoint:
-        return
-    os.environ["S3_ENDPOINT_URL"] = ""
-
-    def cleanup() -> None:
-        if endpoint is not None:
-            os.environ["S3_ENDPOINT_URL"] = endpoint
-
-    testcase.addCleanup(cleanup)
+    with patch.dict(
+        os.environ,
+        {
+            "AWS_ACCESS_KEY_ID": "test-access-key",
+            "AWS_SECRET_ACCESS_KEY": "test-secret-access-key",
+        },
+    ) as patched_environ:
+        for var in (
+            "S3_ENDPOINT_URL",
+            "AWS_SECURITY_TOKEN",
+            "AWS_SESSION_TOKEN",
+            "AWS_PROFILE",
+            "AWS_SHARED_CREDENTIALS_FILE",
+            "AWS_CONFIG_FILE",
+        ):
+            patched_environ.pop(var, None)
+        _get_s3_client.cache_clear()
```

Comment:
It might be worth adding a comment hear that you are clearing the lru cache and not clearing an s3 client cache...